### PR TITLE
koord-manager: enhance batch resource configuration and updating

### PR DESCRIPTION
**Pull request by saintube**
### Ⅰ. Describe what this PR does

koord-manager: enhance batch resource configuration and updating:

1. Refactor the NRT updating in the BatchResource plugin. Add the PreUpdate stage for checking and updating additional objects alongside the framework's Node updating.
2. Support the `cpuCalculatePolicy` in the ColocationStrategy. Enable calculating `batch-cpu` according to high-priority pods' maximal of requests and usages. The default policy is `usage`, which allows the low-priority pods to reclaim the resources requested but unused by the high-priority pods. The new policy `maxUsageAndRequest` is helpful when the cpu resources are not strongly expected to be overcommitted between different priority bands, where neither used resources nor requested resources can be allocatable to the low-priority pods.
3. Support colocation strategy based on the node metadata. Add an annotation protocol for node-level configuration. Add two label protocols for label-preferred use cases.

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/slo-controller/noderesource/framework/extender_plugin.go [hunk 1]
```diff
@@ -24,8 +24,17 @@ import (
...
+//     node resource reconcile loop:
+//  1. Calculate / Reset (determined by whether the config disables node resources, e.g. colocation disabled)
+//  2. NodePreUpdate
+//  3. NodePrepare: (it should be idempotent since it can be called multiple times in one loop)
+//  4. NodeSync
+//  5. NodeMetaSync
 var (
 	globalSetupExtender             = NewRegistry("Setup")
+	globalNodePreUpdateExtender     = NewRegistry("NodePreUpdate")
 	globalNodePrepareExtender       = NewRegistry("NodePrepare")
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/framework/extender_plugin.go:38]: better to rename NodePreparePlugin.Execute to NodePreparePlugin.Prepare

File: pkg/slo-controller/noderesource/framework/extender_plugin.go [hunk 2]
```diff
@@ -24,8 +24,17 @@ import (
 	"github.com/koordinator-sh/koordinator/pkg/slo-controller/metrics"
 )
 
+// node resource initialization:
+//  1. Setup
+//     node resource reconcile loop:
+//  1. Calculate / Reset (determined by whether the config disables node resources, e.g. colocation disabled)
+//  2. NodePreUpdate
+//  3. NodePrepare: (it should be idempotent since it can be called multiple times in one loop)
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/framework/extender_plugin.go:32]: NodeStatusPrepare
or just NodePrepare if some label is included.

File: pkg/slo-controller/noderesource/framework/extender_plugin.go [hunk 3]
```diff
@@ -24,8 +24,17 @@ import (
...
+//  4. NodeSync
+//  5. NodeMetaSync
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/framework/extender_plugin.go:34]: only for need sync checking?
NodeStatusCheck
NodeMetaCheck

File: pkg/slo-controller/noderesource/framework/extender_plugin.go [hunk 4]
```diff
@@ -24,8 +24,17 @@ import (
 	"github.com/koordinator-sh/koordinator/pkg/slo-controller/metrics"
 )
 
+// node resource initialization:
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/framework/extender_plugin.go:27]: need more explanation about the running order of these plugins

e.g. there are several stages:
Setup, Reset if node not enabled

Setup
Calculate(run plugins for generate internal nr)
Update(run PreUpdatePlugin for nr injection or other crd update; run checkNeedSync plugin, run Prepare plugins, do update node obj)

File: apis/extension/node_colocation.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+// NOTE: functions in this file can be overwritten for extension
+
```
Comment (eahydra) [inline apis/extension/node_colocation.go:20]: Maybe we should add some comments to describe what the labels means.

Comment (saintube) [reply apis/extension/node_colocation.go]: done

File: pkg/slo-controller/noderesource/framework/extender_plugin.go (hunk 1 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/framework/extender_plugin.go:38]: done

**Comment by hormes** [comment]
> Enable calculating batch-cpu according to high-priority pods' requests instead of usages.

Explain when you need to use request instead of usage?

File: apis/extension/node_colocation.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+package extension
+
+const (
+	// LabelCPUReclaimRatio denotes the CPU reclaim ratio of a node. The value is a float number.
+	// It takes precedence to the CPUReclaimThresholdPercent in the slo-controller-config.
+	LabelCPUReclaimRatio = NodeDomainPrefix + "/cpu-reclaim-ratio"
+	// LabelMemoryReclaimRatio denotes the memory reclaim ratio of a node. The value is a float number.
+	// It takes precedence to the MemoryReclaimThresholdPercent in the slo-controller-config.
+	LabelMemoryReclaimRatio = NodeDomainPrefix + "/memory-reclaim-ratio"
```
Comment (hormes) [inline apis/extension/node_colocation.go:25]: Should these two fields be placed in annotation?

**Comment by saintube** [comment]
> > Enable calculating batch-cpu according to high-priority pods' requests instead of usages.
> 
> Explain when you need to use request instead of usage?

Add code comments for different calculating policies.

File: apis/extension/node_colocation.go (hunk 6 above)
Comment (saintube) [reply apis/extension/node_colocation.go:32]: These label protocols are for some label-preferred scenarios. A standard annotation protocol is newly added.

**Pull request merged**